File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/L._Просеивание_вниз/solution.cpp
#include <vector>
#include <cassert>
#ifdef REMOTE_JUDGE
#include "solution.h"
#endif


int siftDown(std::vector<int>& heap, int idx) {
    /*
    Time Complexity O(H) -> O(logN)
    Memory Complexity O(H) -> O(logN)
    */
    int upperBound = heap.size() - 1;
    int leftIdx = idx * 2;
    int rightIdx = idx * 2 + 1;
    if (leftIdx > upperBound) return idx;
    int largestIdx = leftIdx;
    if (rightIdx <= upperBound && heap[leftIdx] < heap[rightIdx]) {
        largestIdx = rightIdx;
    }
    if (heap[idx] < heap[largestIdx]) {
        std::swap(heap[idx], heap[largestIdx]);
        return siftDown(heap, largestIdx);
    }
    return idx;
}

#ifndef REMOTE_JUDGE
void test() {
    std::vector<int> sample = {-1, 12, 1, 8, 3, 4, 7};
    assert(siftDown(sample, 2) == 5);
}


int main() {
    test();
}
#endif

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/L._Просеивание_вниз/solution.py
def sift_down(heap, idx) -> int:
    """
    Time Complexity O(H) -> O(logN)
    Memory Complexity O(H) -> O(logN)
    """
    upper_bound: int = len(heap) - 1
    left_idx: int = idx * 2
    right_idx: int = idx * 2 + 1
    if left_idx > upper_bound:
        return idx
    largest_idx: int = left_idx
    if right_idx <= upper_bound and heap[left_idx] < heap[right_idx]:
        largest_idx = right_idx
    if heap[idx] < heap[largest_idx]:
        heap[idx], heap[largest_idx] = heap[largest_idx], heap[idx]
        return sift_down(heap, largest_idx)
    return idx


def test():
    sample = [-1, 12, 1, 8, 3, 4, 7]
    assert sift_down(sample, 2) == 5


if __name__ == '__main__':
    test()